libs/src/api/serviceProvider.ts

```diff
@@ -122,34 +130,53 @@ class ServiceProvider extends Base {
    * @param {any[]} discoveryNodes the verbose list of discovery nodes to select from
    * @param {(node: { delegateOwnerWallet: string }) => boolean} filter an optional filter step to remove certain nodes
    */
-  async getUniquelyOwnedDiscoveryNodes ({ quorumSize, discoveryNodes = [], filter = (node) => true, useWhitelist = true }) {
+  async getUniquelyOwnedDiscoveryNodes({
+    quorumSize,
+    discoveryNodes = [],
+    filter = (_) => true,
+    useWhitelist = true
+  }: {
+    quorumSize: number
+    discoveryNodes?: ServiceWithEndpoint[]
+    filter: (node: ServiceWithEndpoint) => boolean
+    useWhitelist?: boolean
+  }) {
     if (!discoveryNodes || discoveryNodes.length === 0) {
       // Whitelist logic: if useWhitelist is false, pass in null to override internal whitelist logic; if true, pass in undefined
       // so service selector uses internal whitelist
-      discoveryNodes = await this.discoveryProvider.serviceSelector.findAll({ verbose: true, whitelist: useWhitelist ? undefined : null })
+      discoveryNodes = (await this.discoveryProvider.serviceSelector.findAll({
+        verbose: true,
+        whitelist: useWhitelist ? undefined : null
+      })) as ServiceWithEndpoint[]
     }
 
     discoveryNodes.filter(filter)
 
     // Group nodes by owner
-    const grouped = discoveryNodes.reduce((acc, curr) => {
+    const grouped = discoveryNodes.reduce<{
+      [owner: string]: ServiceWithEndpoint[]
+    }>((acc, curr) => {
       if (curr.owner in acc) {
+        // @ts-expect-error
         acc[curr.owner].push(curr)
       } else {
         acc[curr.owner] = [curr]
       }
       return acc
     }, {})
 
-    if (Object.keys(grouped) < quorumSize) {
+    if (Object.keys(grouped).length < quorumSize) {
       throw new Error('Not enough unique owners to choose from')
     }
 
     // Select quorumSize owners from the groups
     const owners = sampleSize(Object.keys(grouped), quorumSize)
 
     // Select 1 node from each owner selected
-    return owners.map(owner => sampleSize(grouped[owner], 1)[0].endpoint)
+    return owners.map(
+      (owner) =>
+        (sampleSize(grouped[owner], 1)[0] as ServiceWithEndpoint).endpoint
```

Comment:
guessing not much we can do with the types here?

Reply:
Nothing simple - it's because indexing into grouped with owner returns potentially undefined

Reply:
I see could we do a ?. And then removeNullable type filter? Or just not worth it